LedgerDB: prune on garbage collection instead of on every change

amesgen opened the pull request:
This is in preparation for #1424

This PR is intended to be reviewed commit-by-commit.

Currently, we prune the LedgerDB (ie remove all but the last `k+1` states) every time we adopt a longer chain. This means that we can not rely on the fact that other threads (like the `copyAndSnapshot` ChainDB background) actually observe all immutable ledger states, just as described in the caveats of our `Watcher` abstraction.

However, a predictable ledger snapshotting rule (#1424) requires this property; otherwise, when the node is under high load and/or we are adopting multiple blocks in quick succession, the node might not be able to create a snapshot for its desired block.

This PR changes this fact: Now, when adopting new blocks, the LedgerDB is *not* immediately pruned. Instead, the a new dedicated background thread for ledger maintenance tasks (flushing/snapshotting/garbage collection) in the ChainDB will periodically (on every new immutable block) wake up and (in particular) garbage collect the LedgerDB based on a slot number.

Also, this makes the semantics more consistent with the existing garbage collection of previously-applied blocks in the LedgerDB, and also with how the ChainDB works, where we also don't immediately delete blocks from the VolatileDB once they are buried beneath `k+1` blocks.

See https://github.com/IntersectMBO/ouroboros-consensus/pull/1513#issuecomment-2944658608 for benchmarks demonstrating that the peak memory usage does not increase while syncing (where we now briefly might hold more than `k+1` ledger states in memory).

jasagredo commented on ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/V2.hs:
```diff
@@ -207,8 +210,9 @@ mkInternals bss h =
-          Right frk ->
+          Right frk -> do
             forkerPush frk st >> atomically (forkerCommit frk) >> forkerClose frk
+            getEnv h pruneLedgerSeq
```
I don't see why we are still pruning in here? I thought we would only prune at garbage collection, which we do below. Couldn't a snapshot be started in between this line and the previous one and then we would close the handle while taking the snapshot?

jasagredo replied:
Ah wait, if we commit the forker, and we try to take a snapshot, then we will take the snapshot of the committed imm tip, which will not be pruned ever

jasagredo replied:
Still this is confusing, because we prune on every change it seems, and the title of the PR says precisely that we prune on GC only.

jasagredo replied:
Ah nevermind. This is part of the testing internals, I was missing that.

jasagredo reviewed (approved):
Looks good.

jasagredo commented on ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/API.hs:
```diff
@@ -775,10 +788,11 @@ type LedgerSupportsLedgerDB blk =
+    LedgerDbPruneAll
+  | -- | Prune to only keep the last @k@ states.
+    LedgerDbPruneKeeping SecurityParam
+  | -- | Prune such that all (non-anchor) states are older than the given slot.
```
```suggestion
  | -- | Prune such that all (non-anchor) states are younger than the given slot.
```

amesgen replied:
Thanks for pointing out this inversion in a couple of places!

Changed, using "not older" instead of "younger" to account for the case of equality (we want to keep states with the same slot as the argument slot).

jasagredo commented on ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/V1/DbChangelog.hs:
```diff
@@ -47,8 +47,8 @@
--- hold the last \(k\) in-memory ledger states. This data type is impemented
--- using the /finger tree/ data structure and has the following time
+-- hold (at least) the last \(k\) in-memory ledger states. This data type is
+-- impemented using the /finger tree/ data structure and has the following time
```
```suggestion
-- implemented using the /finger tree/ data structure and has the following time
```

jasagredo commented on ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/V1/DbChangelog.hs:
```diff
@@ -402,6 +399,14 @@ prune LedgerDbPruneAll dblog =
+ where
+  DbChangelog{changelogStates} = dblog
+
+  -- The anchor of @vol'@ might still have a tip slot larger than @slot@, which
```
```suggestion
  -- The anchor of @vol'@ might still have a tip slot smaller than @slot@, which
```

jasagredo replied:
or `older`

jasagredo commented on ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/V2/LedgerSeq.hs:
```diff
@@ -255,6 +256,12 @@ prune howToPrune (LedgerSeq ldb) = case howToPrune of
+  LedgerDbPruneBeforeSlot slot ->
+    (closeButHead before, LedgerSeq after)
+   where
+    -- The anchor of @vol'@ might still have a tip slot larger than @slot@,
```
```suggestion
    -- The anchor of @vol'@ might still have a tip slot older than @slot@,
```

geo2a commented on ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/ChainDB/Impl/Background.hs:
```diff
@@ -265,8 +270,12 @@ copyAndSnapshotRunner cdb@CDB{..} gcSchedule replayed fuse = do
 
+    -- See the Haddocks above as for why we garbage-collect the LedgerDB already
+    -- here (instead of as part of the scheduled GC).
+    whenJust (withOriginToMaybe gcSlotNo) $ LedgerDB.garbageCollect cdbLedgerDB
```
oops, now I see that chainging this is the entire point of e4143672b!! Please disregard my previous comment.

amesgen commented on ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Storage/LedgerDB/API.hs:
```diff
@@ -251,9 +251,15 @@ data LedgerDB m l blk = LedgerDB
-  -- ^ Garbage collect references to old blocks that have been previously
-  -- applied and committed.
+  , garbageCollect :: SlotNo -> m ()
+  -- ^ Garbage collect references to old state that is older than the given
```
"State" was intended to be a generic term here, as garbage collection does not just clear old ledger states, but also old previously-applied points, and it feels slightly weird to describe these as "states".

amesgen commented:
Thanks for the great reviews, I hope I addressed your comments. Interesting changes:

 - d627e843fd712e3aa725c1843b65939b7bea6295 for removing `LedgerDbPruneKeeping` prompted by https://github.com/IntersectMBO/ouroboros-consensus/pull/1513#discussion_r2179421993
 - Found a bug in the LedgerDB state machine test (#1576) which this PR now depends on as I enriched it due to https://github.com/IntersectMBO/ouroboros-consensus/pull/1513#discussion_r2174925570

amesgen commented on ouroboros-consensus/changelog.d/20250626_193647_alexander.esgen_ledgerdb_garbage_collect_states.md:
```diff
@@ -0,0 +1,6 @@
+### Breaking
+
+- Changed pruning of immutable ledger states to happen on LedgerDB/ChainDB
```
FTR we discussed this in the sync today, and decided that a new ChainDB background thread makes sense.
